File: Days 71 - 80/Day 75/GetLengthOfLongestIncreasingSubsequence.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblem
{
	internal class Day75
	{
		private static Dictionary<int, int> numberCache = new Dictionary<int, int>();

		private static int Main(string[] args)
		{
			int[] array = { 0, 8, 4, 12, 2, 10, 6, 14, 1, 9, 5, 13, 3, 11, 7, 15 };

			Console.WriteLine(GetLongestIncreasingSubsequenceLength(array));

			Console.ReadLine();

			return 0;
		}

		private static int GetLongestIncreasingSubsequenceLength(int[] array) => GetLongestIncreasingSubsequencelengthHelper(array, 0);

		private static int GetLongestIncreasingSubsequencelengthHelper(int[] array, int startIndex)
		{
			if (startIndex == array.Length)
			{
				return 0;
			}

			int currentNumber = array[startIndex];
			int maxIncrement = 1;

			for (int i = startIndex + 1; i < array.Length; i++)
			{
				int count;

				if (array[i] >= currentNumber)
				{
					if (numberCache.ContainsKey(i))
					{
						count = numberCache[i];
					}
					else
					{
						count = GetLongestIncreasingSubsequencelengthHelper(array, i) + 1;
						numberCache.Add(i, count);
					}

					maxIncrement = Math.Max(maxIncrement, count);
				}
			}

			return maxIncrement;
		}
	}
}

File: Days 071 - 080/Day 75/GetLengthOfLongestIncreasingSubsequence.cpp
#include <array>
#include <iostream>
#include <unordered_map>

template <unsigned int Size>
unsigned int GetLongestIncreasingSubsequenceLengthHelper(const std::array<int, Size>& array, unsigned int startIndex) noexcept
{
	static std::unordered_map<unsigned int, unsigned int> numberCache;

	if (startIndex == Size)
	{
		return 0;
	}

	const int currentNumber = array[startIndex];
	unsigned int maxIncrement = 1;

	for (unsigned int i = startIndex + 1; i < Size; i++)
	{
		unsigned int count;

		if (array[i] >= currentNumber)
		{
			if (numberCache.find(i) != numberCache.end())
			{
				count = numberCache[i];
			}
			else
			{
				count = GetLongestIncreasingSubsequenceLengthHelper(array, i) + 1;
				numberCache[i] = count;
			}

			if (count > maxIncrement)
			{
				maxIncrement = count;
			}
		}
	}

	return maxIncrement;
}

template <unsigned int Size>
inline unsigned int GetLongestIncreasingSubsequenceLength(const std::array<int, Size>& array) noexcept
{
	return GetLongestIncreasingSubsequenceLengthHelper(array, 0);
}

int main(int argc, char* argv[])
{
	std::cout << GetLongestIncreasingSubsequenceLength<16>({ 0, 8, 4, 12, 2, 10, 6, 14, 1, 9, 5, 13, 3, 11, 7, 15 }) << "\n";

	std::cin.get();

	return 0;
}